addon/adapters/firebase.js

```diff
@@ -418,42 +418,81 @@ export default DS.Adapter.extend(Waitable, {
     });
 
     return new Promise((resolve, reject) => {
-      var savedRelationships = Ember.A();
+
+
+      var relationshipsToSave = [];
+
+
+      // first we remove all relationships data from the serialized record, we backup the
+      // removed data so that we can save it at a later stage.
       snapshot.record.eachRelationship((key, relationship) => {
-        var save;
-        if (relationship.kind === 'hasMany') {
-          if (serializedRecord[key]) {
-            save = this._saveHasManyRelationship(store, typeClass, relationship, serializedRecord[key], recordRef, recordCache);
-            savedRelationships.push(save);
-            // Remove the relationship from the serializedRecord because otherwise we would clobber the entire hasMany
-            delete serializedRecord[key];
-          }
-        } else {
-          if (this.isRelationshipEmbedded(store, typeClass.modelName, relationship) && serializedRecord[key]) {
-            save = this._saveEmbeddedBelongsToRecord(store, typeClass, relationship, serializedRecord[key], recordRef);
-            savedRelationships.push(save);
+        const data = serializedRecord[key];
+        const isEmbedded = this.isRelationshipEmbedded(store, typeClass.modelName, relationship);
+        const hasMany = relationship.kind === 'hasMany';
+        if ( hasMany || isEmbedded ) {
+            if (!Ember.isNone(data)) {
+              relationshipsToSave.push(
+                {
+                  data:data,
+                  relationship:relationship,
+                  isEmbedded:isEmbedded,
+                  hasMany:hasMany
+                }
+              );
+            }
             delete serializedRecord[key];
           }
         }
-      });
-
-      var relationshipsPromise = Ember.RSVP.allSettled(savedRelationships);
+      );
+      var reportError = (errors) => {
+        var error = new Error(`Some errors were encountered while saving ${typeClass} ${snapshot.id}`);
+        error.errors = errors;
+        reject(error);
+      };
+      // now we update the record
       var recordPromise = this._updateRecord(recordRef, serializedRecord);
-
-      Ember.RSVP.hashSettled({relationships: relationshipsPromise, record: recordPromise}).then((promises) => {
-        var rejected = Ember.A(promises.relationships.value).filterBy('state', 'rejected');
-        if (promises.record.state === 'rejected') {
-          rejected.push(promises.record);
-        }
-        // Throw an error if any of the relationships failed to save
-        if (rejected.length !== 0) {
-          var error = new Error(`Some errors were encountered while saving ${typeClass} ${snapshot.id}`);
-          error.errors = Ember.A(rejected).mapBy('reason');
-          reject(error);
-        } else {
-          resolve();
+      recordPromise.then(
+        () => {
+          if (recordPromise.state === 'rejected') {
```

Comment:
this would probably be related to security rules. its a good question.

If we want to support this, we'd need a backup of previous attributes. This could be grabbed from ember-data.